handle http auth

reepush opened the pull request:
Finally got it working for both Chrome and Firefox!

* In emulation mode callback for _mooltipass.device.retrieveCredentials_ didn't execute that's why there was no auth popup.
* Stuck with a problem that it's impossible to get `Authorization` header while we intercept response. But we can create our custom popup and handle http authorization by ourself. I used window.prompt now, will add popup later.
* Firefox doesn't execute callback because of the _asyncBlocking_ setting. It's not supported and I will see the differences between Chrome's _asyncBlocking_ and Firefox's _blocking_.

limpkin commented on .travis.yml:
```diff
@@ -14,45 +14,43 @@ env:
+  - secure: awBbYRgPzhc0bnI5BZa5L9YMSuWPWhLjIIBIbxIDgi4X0Yk/P4b4u6sD7A/jDbsjRa18uZSztV1f8T2v5KRVUMEsC9wQ31UgH9HeFajQ6s3wpILkuqQEc6DLJReL0Ejf/LgPPaq6cjimzYnQFE8naDkPveQ7IeqKqh6XUdI32jAdEsakaCkPe2t206J/nRP/iMl1Sxzcxxvhc8ASSlUt028dhGGj+khaCizLK2icivedm8S0gfr7Mec65UCogcbAzEM3FsHqXq/wDyKmcUMAoOwnnjj7yhB/8p1UQzrMPF1HWc9Y17w9HfS7MA4W7aaCROnOnkTOoEMEJfPUr+DgWd/6z5TpHx7x6KozNhu73RCBu5W7d72vKqfTVzqWBjBhUgs ...
+  - secure: dg2TJOBrNBljrEkKueaQmILwVA48Sz6/w/kgXsr4VUjo1ml44HV264OmlhcZKJtqVoyDbAVrjDd8bLMLE4/RFyWpyYtnrPdfoIEpJ1Z5RA0IunSrLScb40aaJhDxdr6NK7Gp1IeTPPKb12pGtTb+enrMH5ImyQCwVOverYLqpCEvP9rWE14mE08V7N2vUKjtz69/I+B4q4vqSR5fBamz4NW/0djswEhysEDCXp1rAvi1Q/6hVUcjLtErnAUkjYSSPcQ/ESNtbJ3b13V5qp/vxFO0gJBS5NrFYyO8WlYohs9b33YBsf9cWHOpINJNur/HZET2OfSf9xyd2DrYcJw+uDsLu9YMK3T84F0IHrcBpt/Jg9xK4wIE/dzsdvnJBxcKORH ...
+  - secure: fVwYFsZkADhlLb742KRoxIxRw5MH2gAX3Pxa2YQvJLFh0pkkU/61GTVIhf3yReXAwaRcjES7NzPaqCguT7d19Ak7oD5Uw3dO9q/ZEGd7bDUNT2v+VLeECkY+R/ZKTWTyzAA+IQLR3SuqWAU8i8ocq6LNTPSHD5px0h2ofCapHpFUUuKt2lyEonxjdqRgbjexEha3UOHcC6uvQicFE8AfDGHvDgk1gtRQTYKKLG9+eMxNexVZvc5w5jSlWREVTxula/PqI1qmX/QUDmjmTTmS37cxr0AuAeDQwYoO3NPC+n3Cl0mgcQR33n1bdTMFTdkRPQgffSuCoIpF+tP/DI1mnDi8VgVRRbL90TEwPFw45lHzd//ChZ4eI1/3R+gkZ2hBzlf ...
+script: if [[ $TRAVIS_COMMIT_MESSAGE == *"[ci-ext]"* ]]; then mvn install -Dcucumber.options="src/test/resources/mooltipass/automatedTest/features/$TEST_SUITE"; fi ;
```
please don't push your travis.

limpkin replied:
can you remove this travis from your PR?

limpkin commented on chrome_extension/vendor/mooltipass/mcCombinations.js:
```diff
@@ -140,6 +140,15 @@ var extendedPost = {
+	},
+	'nordinary-com.netcup-mail.de': function(form) {
+		return form.find('#submit')
+	},
```
didn't we agree that we'd try another approach than this one?

reepush replied:
I branched from master, that's why these changes are here. Don't mention them.

limpkin commented on chrome_extension/vendor/mooltipass/mcCombinations.js:
```diff
@@ -941,9 +950,17 @@ mcCombinations.prototype.doSubmit = function doSubmit( currentForm ) {
+			if ( window.location.hostname.indexOf( specialCase ) > -1 ) {
+				submitButton = extendedSubmitButton[specialCase](currentForm.element);
+			}
+		}
```
same as above.

limpkin commented on chrome_extension/background/httpauth.js:
```diff
@@ -69,6 +69,14 @@ httpAuth.loginOrShowCredentials = function (logins) {
     }
     // no logins found
     else {
-        httpAuth.callback({});
```
does that mean that if we don't know a given credential, we'll get these prompts on chrome, firefox AND safari?
this is not ok.

reepush replied:
We will show custom popup for user with login and password inputs.

> But we can create our custom popup and handle http authorization by ourself. I used window.prompt now, will add popup later.

limpkin replied:
can't you restrict your fix for firefox only?

reepush replied:
We need custom popup for both Chrome and Firefox. Look at my comment above.

limpkin commented on chrome_extension/vendor/mooltipass/device.js:
```diff
@@ -329,6 +335,22 @@ mooltipass.device.sendCredentialRequestMessageFromQueue = function()
                 }
                 if (background_debug_msg > 3) mpDebug.log("%c Emulation mode: nothing in buffer!", mpDebug.css('00ff00'));
+                
+                if (!credentialsFound) {
```
wait... if we actually find the credentials, don't we return from the function? (line 328/334)
your !credentialsFound is therefore not necessary, right?

reepush replied:
That's right, didn't mention it. Thanks.

reepush commented:
Force pushed to _http-auth_ branch. Please check the changes.

<img width="424" alt="screen shot 2017-07-06 at 20 57 56" src="https://user-images.githubusercontent.com/8476339/27926136-0c88f1de-6290-11e7-8de6-b8ad97f1ec1a.png">

* Added custom overlay for http auth
* Handled Firefox spec
* Made changes to _credentialsFound_ as discussed previously

What we need to handle also:
* Pressing _Cancel_ in popup should stop request
* Show error after submitting invalid credentials
* Also remember new credentials after invalid ones were submitted
* Handle proxy authorization
* Handle multiple http authorization at the same time

Last two issues relates to the fact that I removed a bunch of code from _background/httpauth.js_. For my test case everything works and sufficient. But I think I need to put some of the code back to solve last issues. Anyway I need to understand what's needed and what's not and now it's easier to add code step by step.

limpkin commented on chrome_extension/mooltipass-content.css:
```diff
@@ -904,4 +907,134 @@ input.mp-hover-password {
+
+/* Overlay */
+
+.mp-overlay {
```
could you rename mp-overlay to something more explicit like mp-http-auth-overlay ?

limpkin commented on chrome_extension/background/httpauth.js:
```diff
@@ -1,74 +1,25 @@
-    mooltipass.device.retrieveCredentials(httpAuth.loginOrShowCredentials, {"id": details.tabId}, url, url, true);
-}
-
-httpAuth.loginOrShowCredentials = function (logins) {
```
why did you have to remove all that code?

limpkin commented on chrome_extension/vendor/mooltipass/device.js:
```diff
@@ -10,7 +10,7 @@ mooltipass.device = mooltipass.device || {};
  * Mooltipass emulation vars
  */
-mooltipass.device.emulation_mode = false
+mooltipass.device.emulation_mode = true
```
please leave it to false!

limpkin commented on chrome_extension/mooltipass-content.js:
```diff
@@ -1572,7 +1572,7 @@ cip.rememberCredentials = function(event, usernameField, usernameValue, password
 		}
 
-		var url = mpJQ(this)[0].action;
+		var url = mpJQ(event.target)[0].action;
```
could you explain that change?

reepush replied:
This line suppose to get action attribute of the form, but this evaluates to `cip` instance and not to DOM element. Changed it to `event.target` and now we are able to save injected proxy address. Take a look at this code in commit:

```js
// Set proxy address as action attribute of the form, so mcCombs will
// save credentials for the right url.
if (req.args[0].isProxy) {
  mpJQ('.mp-popup-http-auth__form').attr('action', req.args[0].proxyURL)
}
```

limpkin commented:
it seems to work well... congrats!
still waiting to read CI integration tests though
